### PR TITLE
Pick optimal model when doing early stopping

### PR DESCRIPTION
**Pull request by epurdyf**

### CODE REVIEW

File: tensor2tensor/bin/t2t-trainer [hunk 1]
```diff
@@ -108,6 +109,28 @@ def main(_):
   else:
     run_experiment()
 
+  if FLAGS.eval_early_stopping_steps is not None: 
+    # update the checkpoint so that it points to the best model
+    checkpoint_state = saver.get_checkpoint_state(FLAGS.output_dir)
+    all_checkpoint_paths = list(checkpoint_state.all_model_checkpoint_paths)
+
+    def extract_step(path):
+      """Extract the step number from a checkpoint path"""
```
Comment (cbockman) [inline tensor2tensor/bin/t2t-trainer]: what is going on in the below is opaque, can you give an example input & output of this func?

Comment (cbockman) [reply tensor2tensor/bin/t2t-trainer]: and needs docstring :)

File: tensor2tensor/bin/t2t-trainer [hunk 2]
```diff
@@ -108,6 +109,28 @@ def main(_):
...
+      return int(path[path.rindex('-') + 1:])
+
+    steps = [extract_step(path) for path in all_checkpoint_paths]
```
Comment (cbockman) [inline tensor2tensor/bin/t2t-trainer]: I'm having a lot of trouble following this code, can you add a couple comments?  e.g., what is it trying to do?  what is happening on L123?

File: tensor2tensor/bin/t2t-trainer (hunk 1 above)
Comment (epurdyf) [reply tensor2tensor/bin/t2t-trainer]: DOne.

File: tensor2tensor/bin/t2t-trainer (hunk 2 above)
Comment (epurdyf) [reply tensor2tensor/bin/t2t-trainer]: Done.

File: tensor2tensor/bin/t2t-trainer [hunk 3]
```diff
@@ -108,6 +109,43 @@ def main(_):
...
+          step: the step number, e.g., 17
+      """
+      return int(path[path.rindex('-') + 1:])
+
+    # get available step numbers
+    steps = [extract_step(path) for path in all_checkpoint_paths]
+
+    # the step we want is the last one that would have allowed us to
+    # stop when we did (at steps[-1])
+    thresh = steps[-1] - FLAGS.eval_early_stopping_steps
```
Comment (cbockman) [inline tensor2tensor/bin/t2t-trainer]: is all_checkpoint_paths absolutely guaranteed to be sorted in the way we want it to be?

we probably should just sort(steps) to ensure it is

File: tensor2tensor/bin/t2t-trainer [hunk 4]
```diff
@@ -108,6 +109,43 @@ def main(_):
...
+
+    # get the last step that is <= thresh
```
Comment (cbockman) [inline tensor2tensor/bin/t2t-trainer]: why wouldn't checkpoint-[thresh] always exist?  I probably don't quite understand the checkpoint saving behavior

Comment (epurdyf) [reply tensor2tensor/bin/t2t-trainer]: The early stopping flags calculate patience based on step number, not iteration number.

File: tensor2tensor/bin/t2t-trainer (hunk 3 above)
Comment (epurdyf) [reply tensor2tensor/bin/t2t-trainer]: Done.

File: tensor2tensor/bin/t2t-trainer (hunk 4 above)
Comment (cbockman) [reply tensor2tensor/bin/t2t-trainer]: 👍 

can you echo that into the code here as a comment?

Comment (cbockman) [reply tensor2tensor/bin/t2t-trainer]: I'd also suggest grabbing not the last step before, but the step closest to thresh (either before or after)

Comment (epurdyf) [reply tensor2tensor/bin/t2t-trainer]: Added a comment. It doesn't make sense to grab the step after the thresh - we've run eval on all the checkpoints being considered, and the one we've chosen is the one that scored the highest.

**Pull request merged**